### ISSUE TITLE
Delta Lake is not reading few parquet files of an application resulting in count mismatch

### ISSUE THREAD

**Issue by sudhikul** [delta-io/delta#489]
Hi, 

I need one urgent help here.  I am using Delta Lake provided by Databricks for storing the staged data from source application. 

When Apache Spark processes the data, the data from source is staged in form of .parquet files and the transaction log directory **_delta_log** is updated with the location of .parquet files in a .json file.  

However, I have observed that, even though an application gets its data in 4 .parquet files and if the json files have reference of only 3 parquet files,  then there will be a mismatch in the count of records when the data is read using the below commands:

spark>  spark.read.format("delta").load("/path/applicationname")

VS

spark> spark.read.format("parquet").load("/path/applicationname")

**My Questions**

- _Why this issue happens ?  Why the delta lake fails to write the location of .parquet file to transaction log?_ 

- _How to fix this particular issue?_  I have seen that if I change the target path where I will capture the data for the same application and re-process, then 9 out of 10 times the issue gets fixed.  But, I cannot keep changing the target path and that's not a clean solution as well. 

Please let me know if you need any additional information. 

Thanks and Regards, 
Sudhindra

**Comment by sudhikul** [comment]
Hi All,

Can you please help here?  This is a critical issue and I am stuck. 

I have used HDFS and Azure Data Lake Gen 1 file systems before and the above mentioned count mismatch issue didn't happen. However, when I use local file system, this issue is seen.  So, is it something to do with the way Spark processes and stores the data in Delta Lake directory in local file system? 

Please provide your inputs at the earliest. 

Thanks and Regards, 
Sudhindra

**Comment by zsxwing** [comment]
@sudhikul The files in the Delta transaction logs are the source of truth. When you write a Spark job to write files into a Delta table, we will write files directly to the file system. Transient errors may cause task failures and leave incomplete files, and speculative tasks may create duplicate files. They may leave some garbage files on the file system. Delta handles this by only committing file paths created by successful tasks to the transaction logs. If you use `parquet` to read a Delta table, you may read the uncommitted files and the result may be incorrect.

**Comment by sudhikul** [comment]
Hi @zsxwing , Thanks for the insights. 

However, I would like to clarify few more things:

1.  We have not observed any task failures from the Spark logs.  Also, the data is getting written in form of parquet files.  But, the transaction log is not capturing all the parquet files, which is a little strange. 

2.  If I read the data in parquet format using "spark.read.format("parquet")" then it is reading the complete set of parquet files.  

3.  Also,  this issue is often seen when I run Spark jobs to load the data into Delta Lake for a large number of applications in one go, i.e bulk loading, where I will be writing around 2 to 3 million records.   

    Compared to the same, if I simply run Spark job to load the data for only the application for which the issue is seen, by configuring a new directory, then it captures all the data.  

4.  One more interesting observation.  Whenever I write the data into Azure Data Lake File System or HDFS, this count mismatch is not seen.  Only when I am using local file system and writing the data, then I am seeing this count mismatch. 

Can you please let me know how to debug this issue ?  Any logs that we can see from transaction log folder?  How to fix this issue?  I tried vaccum command but it didn't help me much. 

Thanks again for your inputs. 

Thanks and Regards,
Sudhindra

**Comment by zsxwing** [comment]
> * We have not observed any task failures from the Spark logs.  Also, the data is getting written in form of parquet files.  But, the transaction log is not capturing all the parquet files, which is a little strange.

An alternative approach is writing the files to a stage directory and moving the successful ones to the output directory. But some storages such as S3 don't support high performance rename (It's implemented as copy the file to the target place + delete the source file). In addition, this approach may still generate garbage parquet files in the output directory when moving multiple files fails in the middle (For example, your storage system is down).

We are aware that some users may try to use other systems to read a Delta table, so we also provide [vacuum](https://docs.delta.io/latest/delta-utility.html#vacuum) to allow the users to clean up these files before running a different system to read a Delta table. In addition, integration with other systems without running vacuum is also possible. We document them in https://docs.delta.io/latest/integrations.html

> * If I read the data in parquet format using "spark.read.format("parquet")" then it is reading the complete set of parquet files.

It's better to avoid using `format("parquet")` to read a Delta table, as it doesn't understand the Delta transaction logs. It will read the files created by failed tasks.

> 3\. Also,  this issue is often seen when I run Spark jobs to load the data into Delta Lake for a large number of applications in one go, i.e bulk loading, where I will be writing around 2 to 3 million records.

Do you enable Spark speculation (the spark conf `spark.speculation` is set to `true`) which will run another attempt of a task when it detects slow tasks? Processing large amount of data has higher chance to hit a failed task or trigger Spark speculation.

> 4\. One more interesting observation.  Whenever I write the data into Azure Data Lake File System or HDFS, this count mismatch is not seen.  Only when I am using local file system and writing the data, then I am seeing this count mismatch.

Besides the above possible reasons, using the local file system may be a possible cause if you have concurrent writes. The local file system doesn't support [Mutual exclusion](https://github.com/delta-io/delta#requirements-for-underlying-storage-systems). Concurrent writes may overwrite transaction logs and cause this issue as well.

> vaccum command but it didn't help me much.

```VACUUM delta.`<your table path>` RETAIN 0 HOURS``` should delete these garbage files. But you need to ensure that there are no concurrent writes being performed on this table when you run this vacuum command. Otherwise, it would delete some uncommitted files that a concurrent job will commit them later.

**Comment by sudhikul** [comment]
Hi @zsxwing , 

Thanks a ton for your inputs.  When I re-performed the bulk loading using Azure Data Lake Gen 1 filesystem, I didn't observe any count mismatches. 

So, can we infer that bulk loading using local file system will not work? 

Please let me know your thoughts. 

Thanks and Regards, 
Sudhindra

**Comment by zsxwing** [comment]
> So, can we infer that bulk loading using local file system will not work?

Could you clarify what bulk loading means? Are there multiple concurrent writers?

**Comment by sudhikul** [comment]
Yes you are correct.  There are 2 separate Spark jobs , both trying to write to the same transaction log directory.  And by bulk loading, I meant each of these Spark jobs is handling more than 1 million records belonging to 20 applications each (i.e total 40 applications). 

As mentioned already in the discussion before, if I load 1 or 2 applications having issue in a separate run, then it goes through fine and there will not be any garbage files generated. 

That's the reason why I was checking whether this is an issue with bulk loading. 

Thanks and Regards, 
Sudhindra

**Comment by sudhikul** [comment]
Hi, 

I have observed similar issue while writing the data into the Azure Data Lake file system as well :( 

But, this time, the load is not bulk - but there are concurrent writers here as well.  This is causing lot of problems in our validation process and everytime I need to assign a separate directory to capture the transaction log. 

Can you please help in understanding why this could be happening even in Azure Data Lake file system also - I believe that file system supports concurrent writes. 

Thanks and Regards,
Sudhindra

**Comment by brkyvz** [comment]
Hi @sudhikul Can you please share your code and what you mean by "bulk loading"?
